common/src/com/thoughtworks/go/config/materials/PluggableSCMMaterial.java

```diff
@@ -0,0 +1,315 @@
+/*************************GO-LICENSE-START*********************************
+ * Copyright 2014 ThoughtWorks, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *************************GO-LICENSE-END***********************************/
+
+package com.thoughtworks.go.config.materials;
+
+import com.google.gson.annotations.Expose;
+import com.google.gson.annotations.SerializedName;
+import com.thoughtworks.go.config.CaseInsensitiveString;
+import com.thoughtworks.go.config.PipelineConfig;
+import com.thoughtworks.go.domain.MaterialInstance;
+import com.thoughtworks.go.domain.MaterialRevision;
+import com.thoughtworks.go.domain.config.ConfigurationProperty;
+import com.thoughtworks.go.domain.materials.*;
+import com.thoughtworks.go.domain.materials.scm.PluggableSCMMaterialInstance;
+import com.thoughtworks.go.domain.materials.scm.PluggableSCMMaterialRevision;
+import com.thoughtworks.go.domain.scm.SCM;
+import com.thoughtworks.go.util.StringUtil;
+import com.thoughtworks.go.util.command.EnvironmentVariableContext;
+import com.thoughtworks.go.util.command.ProcessOutputStreamConsumer;
+import com.thoughtworks.go.util.json.JsonHelper;
+import com.thoughtworks.go.util.json.JsonMap;
+
+import java.io.File;
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+
+import static com.thoughtworks.go.util.command.EnvironmentVariableContext.escapeEnvironmentVariable;
+import static java.lang.String.format;
+import static org.apache.commons.lang.StringUtils.isEmpty;
+import static org.apache.commons.lang.StringUtils.upperCase;
+
+public class PluggableSCMMaterial extends AbstractMaterial {
+    public static final String TYPE = "PluggableSCMMaterial";
+
+    private String scmId;
+
+    @Expose
+    @SerializedName("scm")
+    private SCM scmConfig;
```

Comment:
I dint want to use PluggedSCMConfig since Im already using PluggableSCMConfig in Pipeline -> Materials. I have kept it SCMs & SCM on same lines as PackageRepositories & PackageRepository. I dont mind changing it to SCMsConfig & SCMConfig.